Add TensorExpressions::evaluate overload that takes LHS Tensor as an argument

macedo22 opened the pull request:
## Proposed changes

This PR adds and tests an overload of `TensorExpressions::evaluate` that takes the resultant LHS `Tensor` as an argument to be filled by `evaluate`, as opposed to being constructed and returned. The motivation for this PR is to have tensor expressions support computing the components of previously-constructed/allocated `Tensor`s.

The existing `evaluate` is used like:
```
auto L = TensorExpressions::evaluate<ti_a, ti_b>(R(ti_a, ti_b) + S(ti_a, ti_b));
```
The `evaluate` overload added in this PR is instead used like:
```
Tensor<...> L{};
TensorExpressions::evaluate<ti_a, ti_b>(make_not_null(&L), R(ti_a, ti_b) + S(ti_a, ti_b));
```

Existing lower level tests for `TensorExpressions::evaluate` as well as the higher level test in `Test_MixedOperations.cpp` have been extended to now test both overloads.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Add TensorExpressions::evaluate overload that takes LHS Tensor as arg

macedo22 commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -90,18 +90,110 @@ auto evaluate(const T& rhs_te) {
+  // Fill LHS tensor
+  for (size_t i = 0; i < T::size(); i++) {
+    (*lhs_tensor)[i] =
+        rhs_te.template get<typename T::structure,
```
Making `Tensor`'s `structure` alias public was necessary to be able to use it outside the `Tensor` class here. If that change to `Tensor` is unwanted/problematic, let me know, and I can instead factor out this loop into a new small helper function in the `Tensor` class.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -90,18 +90,109 @@ auto evaluate(const T& rhs_te) {
+template <auto&... LhsTensorIndices, typename T, typename TE,
+          Requires<Tensor_detail::is_tensor<T>::value> = nullptr,
+          Requires<std::is_base_of<Expression, TE>::value> = nullptr>
+void evaluate(const gsl::not_null<T*> lhs_tensor, const TE& rhs_te) {
```
what do you think about changing `T` -> `Tensor<X, LhsSymm, LhsIndexList>` (Or whatever the template paremeter order is)? This means you can remove the `Requires<is_tensor<T>>`

macedo22 replied:
Done in the first fixup!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -90,18 +90,109 @@ auto evaluate(const T& rhs_te) {
+ */
+template <auto&... LhsTensorIndices, typename T, typename TE,
+          Requires<Tensor_detail::is_tensor<T>::value> = nullptr,
+          Requires<std::is_base_of<Expression, TE>::value> = nullptr>
```
You could use `std::is_base_of_v<Expression, TE>` (no more `::value`)

macedo22 replied:
Done in the first fixup!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -63,24 +63,24 @@ constexpr bool contains_indices_to_contract(
-auto evaluate(const T& rhs_te) {
+template <auto&... LhsTensorIndices, typename TE,
+          Requires<std::is_base_of<Expression, TE>::value> = nullptr>
+auto evaluate(const TE& rhs_te) {
```
Is it possible to have this `evaluate` construct the tensor and then call the other? That could remove some of the code duplication in the checks. In other pieces of code we usually have the "return by value" function call the "return by not_null" function. This helps keep code duplication down

macedo22 replied:
Yep! Don't know why I didn't just do that in the first place.

Just want to check, then do we also want [this](https://github.com/sxs-collaboration/spectre/blob/d766ddab22b38f9bf8c026cab58666f0805d5e89/src/DataStructures/Tensor/Tensor.hpp#L146) `Tensor` constructor removed? I imagine the answer is "yes," but I didn't want to assume in case it was useful to keep around for some other purpose

macedo22 replied:
i.e. if the first `evaluate` overload just calls the second one and the second one does the filling, the first one doesn't need this constructor anymore

nilsdeppe replied:
I'd say remove the constructor if it's never called. That way people won't accidentally do something weird with the constructor :)

macedo22 replied:
Done in the second fixup!

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -33,14 +34,18 @@ void test_evaluate_rank_1_impl() noexcept {
-  const Tensor<DataType, Symmetry<1>, TensorIndexTypeList> L_a =
+  const Tensor<DataType, Symmetry<1>, TensorIndexTypeList> L_a_returned =
       ::TensorExpressions::evaluate<TensorIndex>(R_a(TensorIndex));
+  Tensor<DataType, Symmetry<1>, TensorIndexTypeList> L_a_filled{};
```
[choosing a test at random]: could you add a test where the passed in `Tensor` has non-owning `DataVector`s? The best example would be:
```cpp
Variables<tmpl::list<Tags::Tempa<...>>> vars{num_grid_points};
::TensorExpressions::evaluate<TensorIndex>(make_not_null(&get<Tags::Tempa<...>>(vars)),
                                             R_a(TensorIndex));
```
`Tags::Tempa` is in `DataStructures/Tags/TempTensor.hpp`. This is to make sure that a very common use case that helps eliminate memory allocations works correctly. The code _looks_ like it'll work, but I can't compile and run the test in my head :P

macedo22 replied:
Yep! Should this be for only the `DataVector` test cases and not `double`?

macedo22 replied:
Gave this a shot in the 3rd fixup commit

nilsdeppe reviewed (changes requested):
Overall, LGTM, some small requests

macedo22 commented:
In the first fixup, I also updated `TE` -> `RhsTE` and `rhs_te` -> `rhs_tensorexpression` to be consistent with the LHS tparams and param. In the second one, I swapped the ordering of the two overload definitions in the file since the non-`void` overload calls the `void` one.

Haven't yet added the new test case in the first two fixups

macedo22 commented:
@nilsdeppe The third fixup adds the same existing `CHECK`s but for `TempTensor` as the LHS tensor for the test cases where `DataType != double`. Not sure if what I added is what was what you were looking for. Let me know!

edit: 4th fixup just fills in the `auto` tensor types to explicit `Tensor` types

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash

macedo22 commented:
Done!

nilsdeppe commented:
@macedo22 this has a conflict and needs to be rebased now :)

macedo22 commented:
Rebased and updated!

The first (and only real) commit includes resolved conflicts in `Evaluate.hpp` - just having `evaluate` fill out the LHS tensor using multi-indices instead of storage indices in the loop.

The 1st fixup commit updates `evaluate` to not check the provided LHS symmetry against the symmetry from `LhsTensorSymmAndIndices`. This is so that users can specify LHS symmetries that might not be preserved by RHS order of operations.

The 2nd fixup commit adds the `TempTensor` version of the `MixedOperations.hpp` test (as was done in the first/only real commit for all of the `evaluate` tests).

nilsdeppe reviewed (commented):
fixups LGTM, please go ahead and squash once @fmahebert has had a chance to look at them :)

fmahebert commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -143,6 +147,13 @@ void evaluate(
  * different ordering, such as \f$L_{ba} = R_{ab}\f$.
  *
+ * The symmetry of the returned LHS Tensor depends on the order of operations in
+ * in the RHS tensor expression, i.e. how the expression is written. If you
```
Delete the extra "in" at the line break

macedo22 replied:
Good catch!

fmahebert reviewed (commented):
LGTM. Please squash, including this typo

macedo22 commented:
Squashed!
